plugins/modules/foreman_operating_system.py

```diff
@@ -192,81 +184,80 @@ def main():
             ptables=dict(type='list'),
             provisioning_templates=dict(type='list'),
             password_hash=dict(choices=['MD5', 'SHA256', 'SHA512']),
-            parameters=dict(type='dict'),
+            parameters=dict(type='list', elements='dict', options=dict(
+                name=dict(required=True),
+                value=dict(type='raw', required=True),
+                parameter_type=dict(default='string', choices=['string', 'boolean', 'integer', 'real', 'array', 'hash', 'yaml', 'json']),
+            )),
             state=dict(default='present', choices=['present', 'present_with_defaults', 'absent']),
         ),
+        name_map=name_map,
     )
 
-    (operating_system_dict, state) = module.parse_params()
+    entity_dict = module.clean_params()
 
     module.connect()
 
-    try:
-        # Try to find the Operating System to work on
-        # name is however not unique, but description is, as well as "<name> <major>[.<minor>]"
-        entity = None
-        # If we have a description, search for it
-        if 'description' in operating_system_dict and operating_system_dict['description'] != '':
-            entity = find_operating_system_by_title(module, title=operating_system_dict['description'], failsafe=True)
-        # If we did not yet find a unique OS, search by name & version
-        if entity is None:
-            search_dict = {'name': operating_system_dict['name']}
-            if 'major' in operating_system_dict:
-                search_dict['major'] = operating_system_dict['major']
-            if 'minor' in operating_system_dict:
-                search_dict['minor'] = operating_system_dict['minor']
-            entities = find_entities(OperatingSystem, **search_dict)
-            if len(entities) == 1:
-                entity = entities[0]
-    except Exception as e:
-        module.fail_json(msg='Failed to find entity: %s ' % e)
+    # Try to find the Operating System to work on
+    # name is however not unique, but description is, as well as "<name> <major>[.<minor>]"
+    entity = None
+    # If we have a description, search for it
+    if 'description' in entity_dict and entity_dict['description'] != '':
+        entity = module.find_resource_by_title('operatingsystems', entity_dict['description'], failsafe=True)
+    # If we did not yet find a unique OS, search by name & version
+    if entity is None:
+        search_parts = ['{}="{}"'.format(key, entity_dict[key]) for key in ('name', 'major', 'minor') if key in entity_dict]
+        entity = module.find_resource('operatingsystems', ','.join(search_parts), failsafe=True)
```

Comment:
would it be more readable to have `search_string = ','.join(search_parts)` as an own line?

Reply:
probably. `',' ` makes for a really bad parameter. ;)

Reply:
I changed it so that we do not even save the list.